rllib/algorithms/algorithm.py

```diff
@@ -1774,17 +1890,74 @@ def import_policy_model_from_h5(
 
     @override(Trainable)
     def save_checkpoint(self, checkpoint_dir: str) -> str:
-        checkpoint_path = os.path.join(
-            checkpoint_dir, "checkpoint-{}".format(self.iteration)
-        )
-        pickle.dump(self.__getstate__(), open(checkpoint_path, "wb"))
+        """Exports AIR Checkpoint to a local directory and returns its directory path.
+
+        The structure of an Algorithm checkpoint dir will be as follows::
+
+            policies/
+                pol_1/
+                    policy_state.pkl
+                pol_2/
+                    policy_state.pkl
+            checkpoint_version.txt
+            state.pkl
 
-        return checkpoint_path
+        Note: `checkpoint_version.txt` contains a version string (e.g. "v0") helping
+        RLlib to remain backward compatible wrt restoring from checkpoints from
+        Ray 2.0 onwards.
+
+        Args:
+            checkpoint_dir: The directory where the checkpoint files will be stored.
+
+        Returns:
+            The path to the created AIR Checkpoint directory.
+        """
+        state = self.__getstate__()
+
+        # Extract policy states from worker state (Policies get their own
+        # checkpoint sub-dirs).
+        # TODO: "state" key inside "worker" should be renamed to "policy_states".
+        policies = {}
+        if hasattr(self, "workers") and isinstance(self.workers, WorkerSet):
+            policies = state["worker"].pop("state", {})
+
+        # Add RLlib checkpoint version.
+        state["__version__"] = "v1"
```

Comment:
introducing a versioning framework is great.
I think to make this versioning thing really work, we would need to write a little framework, so there can be a list of save() and load() functions, and we can easily find the right versions based on the version number, etc.
can just be TODOs though. we will create the actual framework when we get there.